public/app/features/serviceaccounts/CreateTokenModal.tsx

```diff
@@ -0,0 +1,161 @@
+    setIsExpirationDateValid(isValid);
+    setNewTokenExpirationDate(value);
+  };
+
+  const onCloseInternal = () => {
+    setNewTokenName('');
+    setIsWithExpirationDate(false);
+    setNewTokenExpirationDate('');
+    setIsExpirationDateValid(false);
+    onClose();
+  };
+
+  const modalTitle = (
+    <div className={styles.modalHeaderTitle}>
+      <Icon className={styles.modalHeaderIcon} name="key-skeleton-alt" size="lg" />
+      <span>{!token ? 'Add service account token' : 'Service account token created'}</span>
+    </div>
+  );
+
+  return (
+    <Modal isOpen={isOpen} title={modalTitle} onDismiss={onCloseInternal} className={styles.modal}>
+      {!token ? (
+        <>
+          <FieldSet>
+            <Field
+              label="Display name"
+              description="Optional name to easily identify the token"
+              className={styles.modalRow}
+            >
+              <Input
+                name="tokenName"
+                value={newTokenName}
+                onChange={(e) => {
+                  setNewTokenName(e.currentTarget.value);
+                }}
+              />
+            </Field>
+            <RadioButtonGroup
+              className={styles.modalRow}
+              options={expirationOptions}
+              value={isWithExpirationDate}
+              onChange={setIsWithExpirationDate}
+              size="md"
+            />
+            {isWithExpirationDate && (
+              <Field label="Expiration date" className={styles.modalRow}>
+                <DatePickerWithInput onChange={onExpirationDateChange} value={newTokenExpirationDate} placeholder="" />
+              </Field>
+            )}
+          </FieldSet>
+          <Button onClick={() => onCreateToken(newTokenName)} disabled={isWithExpirationDate && !isExpirationDateValid}>
+            Generate token
+          </Button>
+        </>
+      ) : (
+        <>
+          <FieldSet>
+            <Label
+              description="You will not be able to see or generate it again. Loosing a token requires creating new one."
+              className={styles.modalRow}
+            >
+              Copy the token. It will be showed only once.
+            </Label>
+            <Field label="Token" className={styles.modalRow}>
+              <div className={styles.modalTokenRow}>
+                <Input name="tokenValue" value={token} readOnly />
+                <ClipboardButton
+                  className={styles.modalCopyToClipboardButton}
+                  variant="secondary"
+                  size="md"
+                  getText={() => token}
+                >
+                  <Icon name="copy" /> Copy to clipboard
+                </ClipboardButton>
+              </div>
+            </Field>
+          </FieldSet>
+          <HorizontalGroup>
+            <ClipboardButton variant="primary" getText={() => token} onClipboardCopy={onCloseInternal}>
+              Copy to clipboard and close
+            </ClipboardButton>
```

Comment:
Instead of two buttons - can we just have a behavior so that when the user clicks `Generate token` it's automatically copied to the clipboard and info about it being copied is shown? I'm not sure if the same action should be possible to perform using two separate buttons.

Reply:
That's a good idea. We'll discuss it with @lukasztyrala . Here I just basically implemented existing design.